File: week15/juyoung/problem5/problem5.cpp
#include <string>
#include <iostream>
using namespace std;

string solve(string v) {
	// 1. 입력이 빈 문자열인 경우, 빈 문자열을 반환한다.
	if (v == "") return "";

	// 2. 문자열을 두 "균형잡힌 괄호 문자열" u, v로 분리한다.
	// 2-1. u에 더 이상 분리할 수 없는 균형잡힌 괄호 문자열을 넣는다.
	string u;
	int left = 0, right = 0;
	for (int i = 0; i < v.length(); i++) {
		if (v[i] == '(') {
			left++;
			u += '(';
		}
		else {
			right++;
			u += ')';
		}
		if (left == right) break;
	}

	// 2-2. v에서 u에 포함되는 부분들을 지운다.
	v.erase(0, left + right);

	// 3. 문자열 u가 올바른 괄호 문자열인지 검사한다.
	int flag = 0;
	for (int i = 0; i < u.length(); i++) {
		if (u[i] == '(') flag++;
		else {
			flag--;
			if (flag < 0) break;
		}
	}

	// 3-1. 수행한 결과 문자열을 u에 이어 붙인 후 반환한다.
	if (flag == 0) return u + solve(v);
	else {
		// 4. u가 올바른 문자열이 아니라면
		// 4-1 ~ 4-3.
		v = "(" + solve(v) + ")";
		// 4-4. u에 대한 처리
		u = u.substr(1, u.length() - 2);
		for (int i = 0; i < u.length(); i++) {
			if (u[i] == '(')
				u[i] = ')';
			else u[i] = '(';
		}

		return v + u;
	}
}

string solution(string p) {
	string answer = "";
	answer = solve(p);
	return answer;
}

int main() {
	string answer;
	string input;

	cin >> input;
	answer = solution(input);
	cout << answer;
	
	return 0;
}

/*
(()())()
result: (()())()

)(
result: ()

()))((()
result: ()(())()
*/

File: week16/sangmin/problem5/Problem5.java
package week16.sangmin.problem5;

import java.util.*;

class Solution {

    Map<String, List<Integer>> map = new HashMap<>();

    int binarySearch(String key, int target){
        List<Integer> list = map.get(key);
        int left = 0;
        int right = list.size() - 1;

        while (left <= right){
            int middle = (left + right) / 2;
            if(list.get(middle) < target)
                left = middle + 1;
            else
                right = middle - 1;
        }

        return list.size() - left;
    }

    public int[] solution(String[] info, String[] query) {
        int[] answer = new int[query.length];
        for(String line : info){
            String[] token = line.split(" ");

            String[] langs = new String[]{token[0], "-"};
            String[] fields = new String[]{token[1], "-"};
            String[] careers = new String[]{token[2], "-"};
            String[] soulfoods = new String[]{token[3], "-"};
            for(String lang : langs)
                for(String field : fields)
                    for(String career : careers)
                        for(String soulfood : soulfoods){
                            if(map.containsKey(lang+field+career+soulfood)){
                                map.get(lang+field+career+soulfood).add(Integer.parseInt(token[4]));
                            }else{
                                map.put(lang+field+career+soulfood, new ArrayList<>());
                                map.get(lang+field+career+soulfood).add(Integer.parseInt(token[4]));
                            }
                        }
        }

        for(String key : map.keySet()){
            Collections.sort(map.get(key));
        }


        for(int i = 0; i < query.length; i++){
            String q = query[i].replaceAll(" and ", "");
            int idx = q.indexOf(" ");
            int result = 0;
            if(map.containsKey(q.substring(0, idx)))
                result = binarySearch(q.substring(0, idx), Integer.parseInt(q.substring(idx + 1)));
            answer[i] = result <= 0 ? 0 : result;
        }
        return answer;
    }
}

public class Problem5 {
    public static void main(String[] args) {
        new Solution().solution(new String[]{"java backend junior pizza 150","python frontend senior chicken 210","python frontend senior chicken 150","cpp backend senior pizza 260","java backend junior chicken 80","python backend senior chicken 50"},
                new String[]{"java and backend and junior and pizza 100","python and frontend and senior and chicken 200","cpp and - and senior and pizza 250","- and backend and senior and - 150","- and - and - and chicken 100","- and - and - and - 150"});
    }
}
